netman/adapters/switches/cached.py

```diff
@@ -0,0 +1,321 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from collections import OrderedDict
+
+from netman.core.objects import port_modes
+from netman.core.objects.bond import Bond
+from netman.core.objects.interface import Interface
+from netman.core.objects.switch_base import SwitchBase
+from netman.core.objects.vlan import Vlan
+from netman.core.objects.vrrp_group import VrrpGroup
+
+
+__all__ = ['CachedSwitch']
+
+
+class Cache():
+    object_type = None
+    object_key = None
+    object_extra = {}
+
+    def __init__(self, *args, **kwargs):
+        self.need_refresh = set()
+        self.dict = OrderedDict(*args, **kwargs)
+
+    def __getitem__(self, item):
+        try:
+            return self.dict[item]
+        except KeyError:
+            # don't create the item, just act as it is present
+            self.need_refresh.add(item)
+            params = {self.object_key: item}
+            params.update({k:(v() if callable(v) else v)
+                           for k, v in self.object_extra.iteritems()})
+            return self.object_type(**params)
+
+    def __setitem__(self, key, value):
+        if not self.dict:
+            self.need_refresh.add(None)
+        self.dict[key] = value
+        try:
+            self.need_refresh.remove(key)
+        except KeyError:
+            pass
+
+    def __contains__(self, item):
+        return item in self.dict
+
+    def __len__(self):
+        return len(self.dict)
+
+    def __delitem__(self, key):
+        del self.dict[key]
+        try:
+            self.need_refresh.remove(key)
+        except KeyError:
+            pass
+
+    def values(self):
+        return self.dict.values()
+
+
+class VlanCache(Cache):
+    object_type = Vlan
+    object_key = 'number'
+
+
+class InterfaceCache(Cache):
+    object_type = Interface
+    object_key = 'name'
+
+
+class BondCache(Cache):
+    object_type = Bond
+    object_key = 'number'
+    object_extra = dict(interface=Interface)
+
+
+class CachedSwitch(SwitchBase):
+    def __init__(self, real_switch):
+        super(CachedSwitch, self).__init__(real_switch.switch_descriptor)
+        self.real_switch = real_switch
+        self.vlans_cache = VlanCache()
+        self.interfaces_cache = InterfaceCache()
+        self.bonds_cache = BondCache()
+
+    def connect(self):
+        return self.real_switch.connect()
+
+    def disconnect(self):
+        return self.real_switch.disconnect()
+
+    def start_transaction(self):
+        return self.real_switch.start_transaction()
+
+    def commit_transaction(self):
+        return self.real_switch.commit_transaction()
+
+    def rollback_transaction(self):
+        return self.real_switch.rollback_transaction()
+
+    def end_transaction(self):
+        return self.real_switch.end_transaction()
+
+    def get_vlans(self):
+        if not self.vlans_cache or self.vlans_cache.need_refresh:
```

Comment:
I test if it contains something, a non empty collection is True, an empty one is False.